src/pymor/algorithms/eigs.py

```diff
@@ -0,0 +1,308 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2020 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.algorithms.gram_schmidt import gram_schmidt
+from pymor.core.logger import getLogger
+from pymor.operators.constructions import IdentityOperator
+from pymor.operators.interface import Operator
+
+
+def eigs(A, E=None, k=3, which='LM', b=None, l=None, maxiter=1000, tol=1e-13, seed=0):
+    """Approximate a few eigenvalues of an |Operator|.
+
+    Computes `k` eigenvalues `w` with corresponding eigenvectors `v` which solve
+    the eigenvalue problem
+
+    .. math::
+        A v_i = w_i v_i
+
+    or the generalized eigenvalue problem
+
+    .. math::
+        A v_i = w_i E v_i
+
+    if `E` is not `None`.
+
+    The implementation is based on Algorithm 4.2 in [RL95]_.
+
+    Parameters
+    ----------
+    A
+        The real |Operator| for which the eigenvalues are to be computed.
+    E
+        The |Operator| which defines the generalized eigenvalue problem.
+    k
+        The number of eigenvalues and eigenvectors which are to be computed.
+    which
+        A string specifying which `k` eigenvalues and eigenvectors to compute:
+
+        - `'LM'`: select eigenvalues with largest magnitude
+        - `'SM'`: select eigenvalues with smallest magnitude
+        - `'LR'`: select eigenvalues with largest real part
+        - `'SR'`: select eigenvalues with smallest real part
+        - `'LI'`: select eigenvalues with largest imaginary part
+        - `'SI'`: select eigenvalues with smallest imaginary part
+    b
+        Initial vector for Arnoldi iteration. Default is a random vector.
+    l
+        The size of the Arnoldi factorization. Default is `min(n - 1, max(2*k + 1, 20))`.
+    maxiter
+        The maximum number of iterations.
+    tol
+        The relative error tolerance for the Ritz estimates.
+    seed
+        Random seed which is used for computing the initial vector for the Arnoldi
+        iteration.
+
+    Returns
+    -------
+    w
+        A 1D |NumPy array| which contains the computed eigenvalues.
+    v
+        A |VectorArray| which contains the computed eigenvectors.
+    """
+
+    logger = getLogger('pymor.algorithms.eigs.eigs')
+
+    assert isinstance(A, Operator) and A.linear
+    assert not A.parametric
+    assert A.source == A.range
+
+    if E is None:
+        E = IdentityOperator(A.source)
+    else:
+        assert isinstance(E, Operator) and E.linear
+        assert not E.parametric
+        assert E.source == E.range
+        assert E.source == A.source
+
+    if b is None:
+        b = A.source.random(seed=seed)
+    else:
+        assert b in A.source
+
+    n = A.source.dim
+
+    if l is None:
+        l = min(n - 1, max(2 * k + 1, 20))
```

Comment:
The 20 appears in [scipy.sparse.linalg.eigs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigs.html) where the `l = min(n - 1, max(2 * k + 1, 20))` is just a heuristic for the `l` parameter. I can give it a name and define it at a visible location, however, I would not make it a parameter of the `eigs` function since `l` is already a parameter.

Reply:
Thats fine.